### PR TITLE
Worldbank x axis

### LINKED ISSUE #91: Width of xAxis for a bar chart too short
From our one day range drawn graphs our X domain only goes to XMax, which will be the bottom-left point of the last bar of the graph - it's not coming up "short", it just doesn't know how wide the bar is after XMax. Our bars' width is dependent on the number of keys of the daily received/sent total messages. The smaller the number of keys the wider the bars. 

(The XMax does not account for any additions made to the barWidth , and the text placement does account for the additional width, which seems to be why we're seeing the problem only with the Axis.)

In order to extend the axis all the way to the end, we'll have to account for the additional barWidth in our domain, or append the additional barWidth another way. That might be possible by adding another half day or so on to our xMax, or by shifting each bar to the left of its X position (-barWidth) instead of +barWidth.

### PR DESCRIPTION
**Pull request by DanielMwendwa**
Closes: #91

### CODE REVIEW

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 1]
```diff
@@ -45,10 +45,18 @@ firebase.auth().onAuthStateChanged(function (user) {
 
 const TIMEFRAME_WEEK = 7;
 const TIMEFRAME_MONTH = 30;
-var chartTimeUnit = "1day";
+const EXTEND_X_AXIS_BY_HOURS = 32
+var chartTimeUnit = "10min";
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js:48]: I like this, but I think it's an unrelated change that should have been in another PR

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 2]
```diff
@@ -45,10 +45,18 @@ firebase.auth().onAuthStateChanged(function (user) {
 
 const TIMEFRAME_WEEK = 7;
 const TIMEFRAME_MONTH = 30;
-var chartTimeUnit = "1day";
+const EXTEND_X_AXIS_BY_HOURS = 32
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: The use of the verb 'extend' makes this read like an action, or method name. As it's really a property, consider something like `X_AXIS_EXTENSION_HOURS`

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 3]
```diff
@@ -45,10 +45,18 @@ firebase.auth().onAuthStateChanged(function (user) {
...
+var chartTimeUnit = "10min";
 var isYLimitReceivedManuallySet = false;
 var isYLimitSentManuallySet = false;
 
+// Date objects inherit from Date.prototype
+// Add the addHours property to the date object
+Date.prototype.addHours= function(h){
+    this.setHours(this.getHours()+h);
+    return this;
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: Where possible, try to write functions that either return a value or mutate state only. Otherwise, if you mutate state and return a value it's really hard to predict which object(s) is being updated, which can lead to subtle bugs. 

So in this case, either call this.setHours and don't return any value, or create a new Date object by copying this one, add the hours to that, and return the copy

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 4]
```diff
@@ -45,10 +45,18 @@ firebase.auth().onAuthStateChanged(function (user) {
...
 const TIMEFRAME_MONTH = 30;
-var chartTimeUnit = "1day";
+const EXTEND_X_AXIS_BY_HOURS = 32
+var chartTimeUnit = "10min";
 var isYLimitReceivedManuallySet = false;
 var isYLimitSentManuallySet = false;
 
+// Date objects inherit from Date.prototype
+// Add the addHours property to the date object
+Date.prototype.addHours= function(h){
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: I recommend you don't modify the prototypes of library functions, particularly the standard library. Although you could probably get away with it here due to the small size of the project, in general it could cause problems - what if two libraries both attempted to add a different addHours for example? One of the implementations would be overwritten causing one of the libraries to break in a way which will be hard to debug.

Instead, create a function addHours(date: Date, hours: number): Date

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 5]
```diff
@@ -183,6 +191,7 @@ const update = (data) => {
 
     // Set x and y scales
     const x = d3.scaleTime().range([0, Width]);
+    const x_axis_failed_sms = d3.scaleTime().range([0, Width]);
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]:
```suggestion
    const x_axis_failed_sms_range = d3.scaleTime().range([0, Width]);
```
or 
```suggestion
    const x_axis_failed_sms_scale = d3.scaleTime().range([0, Width]);
```
for added clarity

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 6]
```diff
@@ -183,6 +191,7 @@ const update = (data) => {
...
     const y_total_received_sms = d3.scaleLinear().range([Height, 0]);
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: Whatever you do with the x_axis, it would be nice to replicated here (e.g. y_axis_total_received_sms_scale).
(At the moment, I can't tell from the name what the type is. Appreciate this is a drive-by comment, possibly on someone else's code, but I think that would help)

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 2 above)
Comment (as2388) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: Also can you add a comment explaining the significance of the number 32?

24 would make sense because it's 1 day, but 32 seems very arbitrary to me.

Comment (as2388) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: I also wonder if a fixed constant is the most appropriate solution here. It has the side effect of extending the x-axis beyond the current day when in the 1 day view. Wouldn't it be better to extend the x-axis to the end of the current day?

**Review by as2388 — CHANGES_REQUESTED** [review]
Thanks @DanielMwendwa. Requesting changes so I can take another quick look after you've addressed my comments

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 2 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: I used 32hours because it resolved the problem when I was working on it but I agree working with 24 because I also think that the data from operators also affect the width of the bars,  therefore the margin of the error was large before since I had not resolved the list of operators.

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 3 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: resolved

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 4 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: resolved

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 2 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: > I also wonder if a fixed constant is the most appropriate solution here. It has the side effect of extending the x-axis beyond the current day when in the 1 day view. Wouldn't it be better to extend the x-axis to the end of the current day?

By adding 24hours it extends the x-axis to the end of the current day since it accommodates the width of the bars

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 1 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js:48]: I didn't intend to make the change on this PR  but since it was approved on imaqal, I thought it was better to leave it there than to uncommit the changes.

**Comment by as2388** [comment]
@DanielMwendwa I think you need to push this branch - seeing comments which are labelled 'resolved' but where the code is unchanged

File: public/media_engagement/styles.css [hunk 7]
```diff
@@ -21,11 +21,14 @@
     background-color: #B6D4E7;
     
 }
-.btn-logout{
+.txt-brown {
+    color: #8d3538; 
+}
+.btn-brown{
```
Comment (as2388) [inline public/media_engagement/styles.css]:
```suggestion
.btn-brown {
```

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 8]
```diff
@@ -45,9 +45,10 @@ firebase.auth().onAuthStateChanged(function (user) {
 
 const TIMEFRAME_WEEK = 7;
 const TIMEFRAME_MONTH = 30;
-var chartTimeUnit = "1day";
+var chartTimeUnit = "10min";
 var isYLimitReceivedManuallySet = false;
 var isYLimitSentManuallySet = false;
+extend_x_max_by_hours = (date) => date.setHours(date.getHours() + 24);
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: This modifies the date and returns it. Please do one or the other

Comment (as2388) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: Could you also change the function to name to something like 'add_one_day_to_date'?

 - By saying extend by 'hours' it sounds like users should be able to specify the number of hours, which they can't. 
 - This function's signature and implementation know nothing of x_max, so I think it's unnecessary to include that in the name. The function becomes much more general purpose with that name change.

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 9]
```diff
@@ -218,20 +228,23 @@ const update = (data) => {
...
-    color_scheme = ["#e6194B", "#f58231", "#3cb44b", "#CCCC00", "#4363d8", "#800000", "#f032e6", "#911eb4"]
+    color_scheme = ["#31cece", "#f58231", "#3cb44b", "#CCCC00", "#4363d8", "#800000", "#f032e6", "#911eb4", "#e6194B"]
     let color = d3.scaleOrdinal(color_scheme);
     let colorReceived = d3.scaleOrdinal(color_scheme).domain(receivedKeys);
     let colorSent = d3.scaleOrdinal(color_scheme).domain(sentKeys);
 
     // set scale domain for failed graph
     y_total_failed_sms.domain([0, d3.max(data, function (d) { return d.total_errored; })]);
+    xMin = d3.min(data, d => new Date(d.day));
+    xMax = d3.max(data, d => extend_x_max_by_hours(new Date(d.day)) )
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]:
```suggestion
    xMax = d3.max(data, d => extend_x_max_by_hours(new Date(d.day)))
```

File: public/media_engagement/worldbank_plr_sms_traffic.js [hunk 10]
```diff
@@ -274,7 +287,7 @@ const update = (data) => {
     //Add the X Axis for the total failed sms graph
     total_failed_sms_graph.append("g")
         .attr("transform", "translate(0," + Height + ")")
-        .call(d3.axisBottom(x)
+        .call(d3.axisBottom(x_axis_failed_sms_range)
```
Comment (as2388) [inline public/media_engagement/worldbank_plr_sms_traffic.js]: Sorry to go round on this again, but on second reading today I think `failed_messages_x_axis_range` would read even better

**Review by as2388 — APPROVED** [review]
Thanks @DanielMwendwa, this makes the graphs look much cleaner/less confusing :)

Approved subject to comments being addressed, then over to Luke

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 3 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: I have written another function to avoid this, here: https://github.com/AfricasVoices/Operations-Dashboard/pull/104/commits/7096dda0f36484443029b955ae713b66b14b0c8d

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 6 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: https://github.com/AfricasVoices/Operations-Dashboard/pull/104/commits/7096dda0f36484443029b955ae713b66b14b0c8d

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 10 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: no worries :)

Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: I will rename attributes for y axis in another branch

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 8 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: @as2388 please take a look on this changes that you have requested

Comment (as2388) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: looks better.
 - Is day of type number or Date? The function name suggests it should be a date but the variable suggests a number - please update the function name or variable name appropriately
 - Does JS provide getDays and setDays functions that allow you to setDays(getDays() +1)? I think that would be more readable

File: .vscode/settings.json [hunk 11]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (as2388) [inline .vscode/settings.json:1]: Please don't commit this. Feel free to add .vscode/ to the .gitignore instead though

Comment (DanielMwendwa) [reply .vscode/settings.json:1]: Thank you, I will do so.

**Comment by lukechurch** [comment]
@DanielMwendwa - please can you address the issues that @as2388 has raised before requesting a review from me? Thanks!

**Review by lukechurch — APPROVED** [review]
LGTM subject to @as2388 changes

File: public/media_engagement/worldbank_plr_sms_traffic.js (hunk 8 above)
Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: Day is of type Date, so the function name is correct. I'll just change the variable name.

Comment (DanielMwendwa) [reply public/media_engagement/worldbank_plr_sms_traffic.js]: Yes it does let me change to days. Thank you @as2388 this has helped a great deal.

**Comment by DanielMwendwa** [comment]
> @DanielMwendwa - please can you address the issues that @as2388 has raised before requesting a review from me? Thanks!

@lukechurch the issue has been resolved.
